File: Engine/Axis/Source/Media/Scene/Native/AxNativeSceneReader.cpp
//---------------------------------------------------------------------------------------------
//	Axis 3D Engine Copyright (C) 2017 Palitri
//
//	This software is released under the MIT License http://palitri.com/axis/license/
//
//	For help and documentation, visit http://palitri.com
//---------------------------------------------------------------------------------------------

#include "AxNativeSceneReader.h"

#include "AxNativeSceneID.h"

#include "..\..\..\Engine\Entities\EntitySet\AxEntitySet.h"

#include "..\..\..\Tools\Streams\AxFileStream.h"

#include "..\..\..\Engine\Axis.h"


AxNativeSceneReader::AxNativeSceneReader(Axis *context)
{
    this->context = context;
}

AxNativeSceneReader::~AxNativeSceneReader(void)
{
}

bool AxNativeSceneReader::Read(AxStream &source)
{
	AxHierarchyStreamReader reader(source);

	long long mainChunkEnd = -1;

    while (reader.ReadChunkHeader())
	{
		switch (reader.chunkId)
		{
			case AxNativeSceneID::Main:
				{
					mainChunkEnd = reader.stream->position + reader.chunkSize;
					break;
				}

			case AxNativeSceneID::Resources:
				{
					break;
				}

			case AxNativeSceneID::ResourceSerializationWrap:
				{
					AxResource *resource = this->context->DeserializeResource(reader.stream);
					if (resource != 0)
					{
						if (resource->typeId == AxSettings::classId)
						{
							delete this->context->settings;
							this->context->settings = (AxSettings*)resource;
						}
						else
						{
							long long originalId = resource->id;
							this->context->AddResource(resource);
							this->idMap.Add(originalId, resource->id);
						}
					}

					break;
				}

			case AxNativeSceneID::RootBranch:
				{
					long long rootId = reader.stream->ReadInt64();
					AxResource *rootResource = this->context->FindResourceX(this->GetMappedResourceId(rootId));

					if (rootResource != 0)
					{
						if (rootResource->resourceType == AxResourceType_EntitySet)
						{
							AxEntitySet *root = (AxEntitySet*)rootResource;

							if (this->context->root != 0)
							{
								for (int i = 0; i < root->references.count; i++)
									this->context->root->references.Add(root->references[i]);

								this->context->RemoveResource(root);
							}
							else
							{
								this->context->root = root;
							}
						}
					}

					break;
				}

			case AxNativeSceneID::RuntimeParameters:
				{
					break;
				}

			default:
			{
				if ((reader.chunkSize < 0) || (reader.stream->position + reader.chunkSize > mainChunkEnd))
					return false;

				reader.SkipChunk();

				break;
			}
		}

		if (reader.stream->position >= mainChunkEnd)
			break;
	}

	this->LinkResources();

	return true;
}

void AxNativeSceneReader::LinkResources()
{
	int resourceCount = this->context->resources.count;
    for (int resourceIndex = 0; resourceIndex < resourceCount; resourceIndex++)
    {
		AxResource *resource = this->context->resources[resourceIndex];

        int numProps = ((AxResource*)resource)->properties.count;
        for (int prop = 0; prop < numProps; prop++)
        {
            AxProperty *parameter = ((AxResource*)resource)->properties[prop];
			if (parameter->IsReference() && (parameter->value != 0))
			{
				AxResource *reference = (AxResource*)parameter->value;
				if (reference->resourceType == AxResourceType_None)
				{
					AxResource *ref = this->context->FindResourceX(this->GetMappedResourceId(reference->id));
					parameter->value = ref;
					delete reference;
				}
			}
        }

		if (resource->resourceType == AxResourceType_EntitySet)
        {
            int numReferences = ((AxEntitySet*)resource)->references.count;
            for (int i = 0; i < numReferences; i++)
            {
				AxResource *reference = (AxResource*)((AxEntitySet*)resource)->references[i];
				if (reference->resourceType == AxResourceType_None)
				{
					AxResource *ref = this->context->FindResourceX(this->GetMappedResourceId(reference->id));
					((AxEntitySet*)resource)->references[i] = ref;
					delete reference;
				}
            }
        }
    }
}

long long AxNativeSceneReader::GetMappedResourceId(long long originalId)
{
	long long newId;
	if (this->idMap.GetValue(originalId, newId))
		return newId;
	else
		return -1;
}

File: Web/Media/Scene/Native/AxNativeSceneReader.js
/**
 * Creates a axis native scene reader object.
 * @param {Axis} context Context to use for the reading. All deserialization will be performed using this context
 * @constructor
 */
function AxNativeFileReader(context)
{
    this.context = context;
    
    this.idMap = new AxDictionary();
}

/**
 * Reads native scene data in Axis' native scene format from a given stream
 * @param {AxStream} source Source stream to read data from
 * @return {Boolean} True if the deserialization was successfull
 */
AxNativeFileReader.prototype.Read = function(source)
{
    var reader = new AxHierarchyStreamReader(source);

    var mainChunkEnd = -1;

    while (reader.ReadChunkHeader())
    {
        switch (reader.chunkId)
        {
            case AxNativeSceneId.Main:
            {
                mainChunkEnd = reader.stream.position + reader.chunkSize;
                break;
            }

            case AxNativeSceneId.Resources:
            {
                break;
            }

            case AxNativeSceneId.ResourceSerializationWrap:
            {
                var resource = this.context.DeserializeResource(reader.stream);
                if (resource !== null)
                {
                    if (resource.typeId === AxSettings.classId)
                    {
                        this.context.settings = resource;                        
                    }
                    else
                    {
                        var originalId = resource.id;
                        this.context.AddResource(resource);
                        this.idMap.Add(originalId, resource.id);
                    }
                }

                break;
            }

            case AxNativeSceneId.RootBranch:
            {
                var rootId = reader.stream.ReadInt64();
                var rootResource = this.context.FindResource_1(this.GetMappedResourceId(rootId));

                if (rootResource !== null)
                {
                    if (rootResource.resourceType === AxResourceType.EntitySet)
                    {
                        root = rootResource;

                        if (this.context.root !== null)
                        {
                            for (var i = 0; i < root.references.count; i++)
                                this.context.root.references.Add(root.references.Get(i));

                            this.context.RemoveResource(root);
                        }
                        else
                        {
                            this.context.root = root;
                        }
                    }
                }

                break;
            }

            case AxNativeSceneId.RuntimeParameters:
            {
                break;
            }
            
            default:
            {
                if ((reader.chunkSize < 0) || (reader.stream.position + reader.chunkSize > mainChunkEnd))
                    return false;

                reader.SkipChunk();

                break;
            }
        }

        if (reader.stream.position >= mainChunkEnd)
            break;
    }

    this.LinkResources();

    return true;
};

/**
 * Private
 * Links temporary resource references made while deserializing and reconstructing the scene
 * This is a helper method and is meant to be called only in a specific context
 */
AxNativeFileReader.prototype.LinkResources = function()
{
    var resourceCount = this.context.resources.count;
    for (var resourceIndex = 0; resourceIndex < resourceCount; resourceIndex++)
    {
        var resource = this.context.resources.Get(resourceIndex);

        var numProps = resource.properties.count;
        for (var prop = 0; prop < numProps; prop++)
        {
            var parameter = resource.properties.Get(prop);
            if (parameter.IsReference() && (parameter.value !== null))
            {
                var reference = parameter.value;
                if (reference.resourceType === AxResourceType.None)
                {
                    var ref = this.context.FindResource_1(this.GetMappedResourceId(reference.id));
                    parameter.value = ref;
                }
            }
        }

        if (resource.resourceType === AxResourceType.EntitySet)
        {
            var numReferences = resource.references.count;
            for (var i = 0; i < numReferences; i++)
            {
                var reference = resource.references.Get(i);
                if (reference.resourceType === AxResourceType.None)
                {
                    var ref = this.context.FindResource_1(this.GetMappedResourceId(reference.id));
                    resource.references.Set(i, ref);
                }
            }
        }
    }
};

/**
 * Private
 * Gets the ID of a temporary resource reference, which was made while deserializing and reconstructing the scene
 * This is a helper method and is meant to be called only in a specific context
 * @param {Integer} originalId description
 * @return {Integer} description
 */
AxNativeFileReader.prototype.GetMappedResourceId = function(originalId)
{
    var newId = new AxReferenceParameter();
    if (this.idMap.GetValue(originalId, newId))
        return newId.value;
    else
        return -1;
};
